presto-jdbc/src/main/java/com/facebook/presto/jdbc/ReferenceCountingLoadingCache.java

```
@@ -0,0 +1,214 @@
+                    {
+                        Holder<V> holder = notification.getValue();
+                        /*
+                         * The docs say that both the key and value may be
+                         * null if they've already been garbage collected.
+                         * We aren't using weak or soft keys or values, so
+                         * this shouldn't apply.
+                         */
+                        requireNonNull(holder, format("holder is null while removing key %s", notification.getKey()));
+                        checkState(holder.getRefcount() == 0, "Non-zero refcount disposing %s", notification.getKey());
+                        checkState(holder.getCleanerCount() == 0, "Non-zero refcount disposing %s", notification.getKey());
+                        disposer.accept(holder.get());
+                    }
+                })
+                .ticker(ticker)
+                .build(new CacheLoader<K, Holder<V>>() {
+                    @Override
+                    public Holder<V> load(K key)
+                            throws Exception
+                    {
+                        return new Holder<>(loader.load(key));
+                    }
+                });
+    }
+
+    public static class Builder<K, V>
+    {
+        Ticker ticker = Ticker.systemTicker();
+
+        public Builder<K, V> ticker(Ticker ticker)
+        {
+            this.ticker = requireNonNull(ticker, "ticker is null");
+            return this;
+        }
+
+        public ReferenceCountingLoadingCache<K, V> build(CacheLoader<K, V> loader, Consumer<V> disposer)
+        {
+            return new ReferenceCountingLoadingCache<K, V>(loader, disposer, ticker);
+        }
+    }
+
+    public static <K, V> Builder<K, V> builder()
+    {
+        return new Builder<>();
+    }
+
+    public void close()
+    {
+        if (closed.compareAndSet(false, true)) {
+            valueCleanupService.shutdownNow();
```

How about failing fast with a checkState(cache.size() == 0)?